lib/Lift.cpp

```diff
@@ -540,59 +541,56 @@ CreateConstFromMemory(const uint64_t addr, llvm::Type *type,
-      auto struct_type = llvm::dyn_cast<llvm::StructType>(type);
-
-      auto num_elms = struct_type->getNumElements();
-      auto elm_offset = 0;
-
-      std::vector<llvm::Constant *> const_list;
+      const auto struct_type = llvm::dyn_cast<llvm::StructType>(type);
+      const auto layout = dl.getStructLayout(struct_type);
+      const auto num_elms = struct_type->getStructNumElements();
+      std::vector<llvm::Constant *> initializer_list;
+      initializer_list.reserve(num_elms);
 
       for (std::uint64_t i = 0U; i < num_elms; ++i) {
-        auto elm_type = struct_type->getElementType(i);
-        auto elm_size = dl.getTypeSizeInBits(elm_type);
-
-        auto const_elm =
-            CreateConstFromMemory(addr + elm_offset, elm_type, arch,
-                                  program, module);
-
-        const_list.push_back(const_elm);
-        elm_offset += elm_size / 8;
+        const auto elm_type = struct_type->getStructElementType(i);
+        const auto offset = layout->getElementOffset(i);
+        auto const_elm = CreateConstFromMemory(addr + offset, elm_type,
+                                               arch, program, module);
+        initializer_list.push_back(const_elm);
       }
-
       result = llvm::ConstantStruct::get(struct_type,
-                                        llvm::ArrayRef(const_list));
+                                         llvm::ArrayRef(initializer_list));
     } break;
 
     case llvm::Type::ArrayTyID: {
+
+      // Traverse through all the elements of array and create the initializer
+      const auto array_type = llvm::dyn_cast<llvm::ArrayType>(type);
       const auto elm_type = type->getArrayElementType();
-      const auto elm_size = dl.getTypeSizeInBits(elm_type);
+      const auto elm_size = dl.getTypeAllocSize(elm_type);
       const auto num_elms = type->getArrayNumElements();
-      std::string bytes(dl.getTypeSizeInBits(type) / 8, '\0');
+      std::vector<llvm::Constant *> initializer_list;
+      initializer_list.reserve(num_elms);
+
       for (auto i = 0u; i < num_elms; ++i) {
-        const auto elm_offset = i * (elm_size / 8);
-        const auto src =
-            ReadValueFromMemory(addr + elm_offset, elm_size, arch, program)
-                .getRawData();
-        const auto dst = bytes.data() + elm_offset;
-        std::memcpy(dst, src, elm_size / 8);
-      }
-      if (elm_size == 8) {
-        result = llvm::ConstantDataArray::getString(module.getContext(), bytes,
-                                                    /*AddNull=*/false);
-      } else {
-        result = llvm::ConstantDataArray::getRaw(bytes, num_elms, elm_type);
+        const auto elm_offset = i * elm_size;
+        auto const_elm = CreateConstFromMemory(addr + elm_offset, elm_type,
+                                               arch, program, module);
+        initializer_list.push_back(const_elm);
       }
+      result = llvm::ConstantArray::get(array_type, initializer_list);
```

Comment:
Will this produce the same output bitcode as before? Rellic migth rely on this to produce string literals later on.

Reply:
The previous implementation was not handling the array of type structs. It should produce the same bitcode for string literals.